src/roadmap/data/app_streams.py

```diff
@@ -7766,4 +7766,15 @@ def set_os_version(self):
     ),
 ]
 
+APP_STREAM_MODULES_BY_KEY = {(asm.name, asm.os_major, asm.stream): asm for asm in APP_STREAM_MODULES}
+
+OS_MAJORS_BY_APP_NAME = {}
+for asm in APP_STREAM_MODULES:
+    try:
+        major_versions = OS_MAJORS_BY_APP_NAME[asm.name]
+    except KeyError:
+        major_versions = set()
+        OS_MAJORS_BY_APP_NAME[asm.name] = major_versions
+    major_versions.add(asm.os_major)
```

Comment:
Creating these different perspectives is a really good idea. It's also probably an indicator we need to rethink the data model, but that's for the future.

This can be simplified using a `defaultdict`
```suggestion
OS_MAJORS_BY_APP_NAME = defaultdict(set)
for asm in APP_STREAM_MODULES:
    OS_MAJORS_BY_APP_NAME[asm.name].add(asm.os_major)
```

Reply:
I didn't make it a defaultdict because I thought it would make it prone to leaking if missing elements were requested.

Reply:
I would prefer to freeze it, but there don't know a lightweight way to do that.

Reply:
@r0x0d I'm curious how it would help here. I read up on dataclasses just now, it seems it can freeze the assignment of a field, but I don't see how it can freeze adding to a dictionary or set.

Reply:
> I didn't make it a defaultdict because I thought it would make it prone to leaking if missing elements were requested.

Ah, I see what you mean. You could change it to a `dict`. Doing all this at the module global level feels kinda weird, but I guess it works.

```suggestion
OS_MAJORS_BY_APP_NAME = defaultdict(set)
for asm in APP_STREAM_MODULES:
    OS_MAJORS_BY_APP_NAME[asm.name].add(asm.os_major)
OS_MAJORS_BY_APP_NAME = dict(OS_MAJORS_BY_APP_NAME)
```

Reply:
A dataclass (or any class) could work by raising an `AttributeError` instead setting a default value when accessing an attribute that doesn't exist.

We could dynamically create a frozen Pydantic model to house the data, but we would have to use `getattr` for items that have a `-` in them.

```python
OSMajorAppName = create_model("OSMajorAppName", **{asm.name: (set[int], ...) for asm in APP_STREAM_MODULES})```

Reply:
If we wanted to keep the global namespace a bit cleaner, we could make function and call that.

```python
def get_os_majors_by_app_name(values: list[AppStreamEntity]) -> dict[str, set[int]]:
    result = defaultdict(set)
    for asm in values:
        result[asm.name].add(asm.os_major)

    return dict(result)

OS_MAJORS_BY_APP_NAME = get_os_majors_by_app_name(APP_STREAM_MODULES)
```

This isn't a blocker, just musing over some different ways to do things.

Reply:
I made a change here. I hear you on the choices available.